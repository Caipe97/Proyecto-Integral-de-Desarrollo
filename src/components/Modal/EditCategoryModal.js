import React, { useState } from 'react';
import Modal  from 'react-bootstrap/Modal';
import { Button } from 'react-bootstrap';
import Combobox from "react-widgets/Combobox";
import "react-widgets/styles.css";
import TextField from '@material-ui/core/TextField';
import Edit from '../../images/edit.png';

const EditCategoryModal = (props) =>{
  const [modalShow, setModalShow] = useState(false);
  let onlyCustomCategories = props.foodCategories.filter((category) => category.userId !== null);
  const [state, setState] = useState({
    name: '',
    foodCategoryId: '',
    categoryName: '',
    message: ''
  })

  const handleChange = event => {
    setState({
      ...state,
      [event.target.name]: event.target.value
    })
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    if(state.name !== '' && state.foodCategoryId !== ''){
      const data = await props.onEditCategory(state.foodCategoryId, state.name);
      if(data){
        setState({
          name: '',
          foodCategoryId: '',
          message: 'Categoría editada exitosamente'
        });
      }
    } else {
      setState({...state, message: 'Debe seleccionar alguna categoría y elegir un nuevo nombre'});
    }
  };

  const getFoodCategoryName = (foodCategoryId) => {
    const foodCategoryName = props.foodCategories.find(foodCategory => foodCategory.foodCategoryId === foodCategoryId);
    if(foodCategoryId === -2){
      return '';
    }
    return foodCategoryName;
  }

  const onChangeComboBox = async event => {
    await setState({...state, foodCategoryId: event.foodCategoryId, categoryName: getFoodCategoryName(event.foodCategoryId)});
  }

  return (
    <>
      <img src={Edit} alt='editar' style={{width: '30px', height: '30px', cursor: 'pointer'}} onClick={() => setModalShow(true)}/>
      <Modal
        show={modalShow}
        size="lg"
        aria-labelledby="contained-modal-title-vcenter"
        centered
      >
        <Modal.Header>
          <Modal.Title id="contained-modal-title-vcenter">
          {'Editar Categoría'}
          </Modal.Title>
          <Button type="button" onClick={() => {setModalShow(false); setState({...state, message: ''});}} style={{ backgroundColor: 'white', borderColor: 'white', color: "black"}}>X</Button>
        </Modal.Header>
        <Modal.Body>

          <form style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100%', flexDirection: 'column' }}>
            <Combobox
              data={onlyCustomCategories}
              value={state.categoryName}
              textField='name'
              onSelect={onChangeComboBox}
              groupBy={category => category.userId}
              renderListGroup={ ({group}) => ( //group es el userId
                <span>{group === null ? 'Default' : 'Custom'}</span>
              )}
            />
            <TextField label="Nuevo nombre" name='name' type='name' value={state.name} onChange={handleChange} required />
            <button onClick={handleSubmit} className='button'>Editar</button>
            <p>{state.message}</p>
            {props.isPending
              ? <div className="spinner-border" role="status">
                  <span className="sr-only"></span>
                </div>
              : null}
          </form>
        </Modal.Body>
      </Modal>
    </>
  );
}
  
  export default EditCategoryModal;