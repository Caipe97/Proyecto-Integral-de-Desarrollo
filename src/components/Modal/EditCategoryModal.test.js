import React from 'react';
import { shallow } from 'enzyme';
import EditCategoryModal from './EditCategoryModal';

describe('ModalEditCategoryModal tests', () => {
    const preventDefault = { preventDefault: jest.fn() }
    let wrapper;
    let mockProps;
    beforeEach(() => {
        mockProps = {
            state: {
                name: '',
                foodCategoryId: '',
                errorMessage: ''
            },
            categories:[
                {
                  foodCategoryId: 1,
                  name: "Fruta",
                  createdAt: "2021-09-30T14:09:34.756Z",
                  updatedAt: "2021-09-30T14:09:34.756Z",
                  userId: null
                },
                {
                  foodCategoryId: 7,
                  name: "Fideos",
                  createdAt: "2021-09-30T18:58:58.542Z",
                  updatedAt: "2021-09-30T18:58:58.542Z",
                  userId: null
                }
              ],
            preventDefault: jest.fn(),
            onEditCategory: jest.fn()
        };
        wrapper = shallow(<EditCategoryModal {...mockProps} />

        );

    })
    it('expect t2', () => {
        expect(wrapper).toMatchSnapshot();
    })
    it('expect to change  onclick buttons', () => {

        wrapper.find('[type="button"]').at(0).simulate('click');
        expect(wrapper.find('div')).toBeDefined();
        expect(wrapper.find('div.p').contains('')).toBeDefined();

    })
    it('expect to change  onclick buttons in img', () => {

        wrapper.find('img').at(0).simulate('click');

    })
    it('handleSubmit expect t11', async () => {

        await wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
    })
   
     it('changecombobox expect to render EditCategoryModal component to location state not in button push profile', async() => {
    
       await  wrapper.find('Combobox').at(0).simulate('onSelect', { target: { name: 'name', value: 'tomate' } });
    
     })
     it('handleChange should update the state correctly when writing on the inputs', () => {
        wrapper.find('[name="name"]').at(0).simulate('change', { target: { name: 'name', value: 'frituras' } });
     
    })
})
