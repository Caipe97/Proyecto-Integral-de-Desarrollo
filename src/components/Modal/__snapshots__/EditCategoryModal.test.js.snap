// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalEditCategoryModal tests expect t2 1`] = `
<Fragment>
  <img
    alt="editar"
    onClick={[Function]}
    src="edit.png"
    style={
      Object {
        "cursor": "pointer",
        "height": "30px",
        "width": "30px",
      }
    }
  />
  <Modal
    animation={true}
    aria-labelledby="contained-modal-title-vcenter"
    autoFocus={true}
    backdrop={true}
    centered={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    enforceFocus={true}
    keyboard={true}
    restoreFocus={true}
    show={false}
    size="lg"
  >
    <ModalHeader
      closeButton={false}
      closeLabel="Close"
    >
      <ModalTitle
        id="contained-modal-title-vcenter"
      >
        Editar Categor√≠a
      </ModalTitle>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "white",
            "color": "black",
          }
        }
        type="button"
        variant="primary"
      >
        X
      </Button>
    </ModalHeader>
    <ModalBody>
      <form
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "center",
          }
        }
      >
        <Combobox
          data={Array []}
          groupBy={[Function]}
          onSelect={[Function]}
          renderListGroup={[Function]}
          textField="name"
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          label="Nuevo nombre"
          name="name"
          onChange={[Function]}
          required={true}
          type="name"
          value=""
        />
        <button
          className="button"
          onClick={[Function]}
        >
          Editar
        </button>
        <p />
      </form>
    </ModalBody>
  </Modal>
</Fragment>
`;
